src/backends.rs

```diff
@@ -123,69 +149,151 @@ cfg_if! {
+                        or Rust version older than 1.80 was used"
+                    );
+                }
             }
         }
     } else if #[cfg(target_os = "hermit")] {
         mod hermit;
-        pub use hermit::*;
+        pub use hermit::Implementation;
     } else if #[cfg(target_os = "vxworks")] {
         mod vxworks;
-        pub use vxworks::*;
+        pub use vxworks::Implementation;
     } else if #[cfg(target_os = "solid_asp3")] {
         mod solid;
-        pub use solid::*;
+        pub use solid::Implementation;
     } else if #[cfg(all(windows, any(target_vendor = "win7", getrandom_windows_legacy)))] {
         mod windows7;
-        pub use windows7::*;
+        pub use windows7::Implementation;
     } else if #[cfg(windows)] {
         mod windows;
-        pub use windows::*;
+        pub use windows::Implementation;
     } else if #[cfg(all(target_arch = "x86_64", target_env = "sgx"))] {
         mod rdrand;
-        pub use rdrand::*;
-    } else if #[cfg(all(target_arch = "wasm32", any(target_os = "unknown", target_os = "none")))] {
-        compile_error!(concat!(
-            "The wasm32-unknown-unknown targets are not supported by default; \
-            you may need to enable the \"wasm_js\" configuration flag. Note \
-            that enabling the `wasm_js` feature flag alone is insufficient. \
-            For more information see: \
-            https://docs.rs/getrandom/", env!("CARGO_PKG_VERSION"), "/#webassembly-support"
-        ));
+        pub use rdrand::Implementation;
     } else {
-        compile_error!(concat!(
-            "target is not supported. You may need to define a custom backend see: \
-            https://docs.rs/getrandom/", env!("CARGO_PKG_VERSION"), "/#custom-backend"
-        ));
+        use_fallback_or! {
+            cfg_if! {
+                if #[cfg(all(target_arch = "wasm32", any(target_os = "unknown", target_os = "none")))] {
+                    compile_error!(concat!(
+                        "The wasm32-unknown-unknown targets are not supported by default; \
+                        you may need to enable the \"wasm_js\" configuration flag. Note \
+                        that enabling the `wasm_js` feature flag alone is insufficient. \
+                        For more information see: \
+                        https://docs.rs/getrandom/", env!("CARGO_PKG_VERSION"), "/#webassembly-support"
+                    ));
+                } else {
+                    compile_error!(concat!(
+                        "target is not supported. You may need to define a custom backend see: \
+                        https://docs.rs/getrandom/", env!("CARGO_PKG_VERSION"), "/#custom-backend"
+                    ));
+                }
+            }
+        }
+    }
+}
+
+/// Provides entropy suitable for CSPRNG.
+///
+/// # Safety
+///
+/// Implementors must uphold the contracts of all methods provided by this trait.
+pub unsafe trait Backend {
```

Comment:
Switching to a trait based backend adds a lot of noise to this PR. I don't have a strong preference etiher for or against, but it should be merged in a separate PR.

I think @newpavlov would prefer not using a trait at all. @bushrat011899 do you have motivation for it?